src/main/java/org/eln2/mc/common/cells/foundation/CellGraph.kt

```diff
@@ -0,0 +1,393 @@
+    val cells = ArrayList<CellBase>()
+
+    private val posCells = HashMap<CellPos, CellBase>()
+
+    private val circuits = ArrayList<Circuit>()
+
+    private val simulationStopLock = ReentrantLock()
+
+    // This is the simulation task. It will be null if the simulation is stopped
+    private var runningTask: ScheduledFuture<*>? = null
+
+    @CrossThreadAccess
+    private var updates = 0L
+
+    private var updatesCheckpoint = 0L
+
+    @CrossThreadAccess
+    var lastTickTime = 0.0
+        private set
+
+    /**
+     * Gets the number of updates that have occurred since the last call to this method.
+     * */
+    fun sampleElapsedUpdates(): Long {
+        val elapsed = updates - updatesCheckpoint
+        updatesCheckpoint += elapsed
+
+        return elapsed
+    }
+
+    /**
+     * True, if the solution was found last simulation tick. Otherwise, false.
+     * */
+    var successful = false
+        private set
+
+    /**
+     * Checks if the simulation is running. Presumably, this is used by logic that wants to mutate the graph.
+     * It also checks if the caller is the server thread.
+     * */
+    private fun validateMutationAccess() {
+        if (runningTask != null) {
+            error("Tried to mutate the simulation while it was running")
+        }
+
+        if (Thread.currentThread() != ServerLifecycleHooks.getCurrentServer().runningThread) {
+            error("Illegal cross-thread access into the cell graph")
+        }
+    }
+
+    /**
+     * Runs one simulation step. This is called from the update thread.
+     * */
+    @CrossThreadAccess
+    private fun update() {
+        simulationStopLock.lock()
+
+        lastTickTime = Time.toSeconds(measureNanoTime {
+            successful = true
+
+            circuits.forEach {
+                successful = successful && it.step(0.05)
+            }
+        })
+
+        updates++
+
+        simulationStopLock.unlock()
+    }
+
+    /**
+     * This realizes the object subsets and creates the underlying simulations.
+     * The simulation must be suspended before calling this method.
+     * @see stopSimulation
+     * */
+    fun buildSolver() {
+        validateMutationAccess()
+
+        cells.forEach { it.clearObjectConnections() }
+        cells.forEach { it.recordObjectConnections() }
+
+        realizeElectrical()
+
+        cells.forEach { it.build() }
+    }
+
+    /**
+     * This method realizes the electrical circuits for all cells that have an electrical object.
+     * */
+    private fun realizeElectrical() {
+        LOGGER.info("Realizing electrical components.")
+
+        circuits.clear()
+
+        realizeComponents(SimulationObjectType.Electrical) { set ->
+            val circuit = Circuit()
+
+            set.forEach { it.electricalObject.setNewCircuit(circuit) }
+
+            circuits.add(circuit)
+
+            LOGGER.info("Found circuit with ${circuit.components.size} components")
+        }
+    }
```

Comment:
I wonder if refactoring the electric sim out of this class would make more sense as we add more simulators.